lightning/src/ln/onchain_utils.rs

```diff
@@ -88,3 +123,649 @@ pub(crate) fn get_witnesses_weight(inputs: &[InputDescriptors]) -> usize {
 	}
 	tx_weight
 }
+
+/// A struct to describe a revoked output and corresponding informations to generate a solving
+/// witness.
+///
+/// If the revoked output is a HTLC output on a committment transaction, HTLCOuputInCommitment (hash
+/// timelock, direction), pubkeys are used to generate the suiting witnessScript.
+///
+/// Otherwise, CSV is used to generate a witnessScript redeeming a balance output on a commitment
+/// transaction or second-stage HTLC transaction.
+///
+/// Amount and revocation secret are used in both cases.
+#[derive(Clone, PartialEq)]
+pub(crate) struct RevokedOutput {
+	per_commitment_point: PublicKey,
+	counterparty_delayed_payment_base_key: PublicKey,
+	counterparty_htlc_base_key: PublicKey,
+	per_commitment_key: SecretKey,
+	input_descriptor: InputDescriptors,
+	amount: u64,
+	htlc: Option<HTLCOutputInCommitment>,
+	on_counterparty_tx_csv: u16,
+}
+
+impl RevokedOutput {
+	pub(crate) fn build(per_commitment_point: PublicKey, per_commitment_key: SecretKey, counterparty_delayed_payment_base_key: PublicKey, counterparty_htlc_base_key: PublicKey, input_descriptor: InputDescriptors, amount: u64, htlc: Option<HTLCOutputInCommitment>, on_counterparty_tx_csv: u16) -> Self {
+		RevokedOutput {
+			per_commitment_point,
+			counterparty_delayed_payment_base_key,
+			counterparty_htlc_base_key,
+			per_commitment_key,
+			input_descriptor,
+			amount,
+			htlc,
+			on_counterparty_tx_csv
+		}
+	}
+}
+
+impl Writeable for RevokedOutput {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
```

Comment:
nit here and below: can we implement this with `impl_writeable!()`?

Reply:
Yep good idea, done.